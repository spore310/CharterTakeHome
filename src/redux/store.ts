import {configureStore} from '@reduxjs/toolkit';
import {useDispatch, useSelector} from 'react-redux';
import movie from './movieSlice';
const store = configureStore({
    reducer:{
        movie
    },
    middleware: (getDefaultMiddleware) => getDefaultMiddleware()
})

export type AppDispatch = typeof store.dispatch
export type RootState = ReturnType<typeof store.getState>


export default store;
