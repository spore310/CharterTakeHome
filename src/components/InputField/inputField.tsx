import React, { FC } from 'react';
import { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { hydrateList, filterList, filterTitle } from '../../redux/movieSlice';
import { RootState, AppDispatch } from '../../redux/store';

import style from './inputField.module.css';
export interface ResponseAxios {
    id: string;
    title: string;
    genres: string[]
}
export const InputField: FC<any> = () => {
    const [input, setInput] = useState<string>("");

    const [showAuto, toogleAuto] = useState<boolean>(false);

    const dispatch = useDispatch<AppDispatch>();

    const myState = (state: RootState) => state.movie;

    const State = useSelector(myState);


    

    useEffect(()=>{
        dispatch(filterTitle(input));
       /*  console.log('input', input);
        let temp = [...State.movieList];
        if(State.genre){
        temp = temp.filter((ele, index, arr)=>{
            return ele.genres.includes(State.genre)
        })
        }
        if(State.keyword){
            temp = temp.filter((ele)=>ele.title.includes(State.keyword))
        }
        console.log('state', temp) */
    },[State.genre,State.keyword, input])



    const handleChange = (e: any) => {
        setInput(e.target.value);
    };

    return (<div className={style.wrapper} data-testid="inputfield-wrapper">
       
       
        <div className={style.inputWrapper}>
            
            <input type='text' value={input} onChange={handleChange} data-testid="inputfield-input"/>
           
            <div className={style.selectorWrapper}>
                <button className={style.genreButton} onClick={()=>toogleAuto(prev=>!prev)}>Genre</button>
                <div className={style.auto}>
                    {showAuto && State.genres && State.genres.map((ele)=>{
                        return <div className={style.autoItem} onClick={()=>{
                            dispatch(filterList(ele))
                            toogleAuto(false);
                        }}>{ele}</div>}
                        )}
                </div>
            {State.genre && <button className={style.genreCancel} onClick={()=>{dispatch(filterList(''))}}>Clear Genre</button>}
            </div>
        </div>
    </div>)
}
