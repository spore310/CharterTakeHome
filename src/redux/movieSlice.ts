import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import { ResponseAxios } from "../components/inputField";
import axios from 'axios';

export const hydrateList = createAsyncThunk('movie/hydrateList', async() => {
    
    const {data:{data}} = await axios.get('https://code-challenge.spectrumtoolbox.com/api/movies',{headers:{
                'Authorization': 'Api-Key q3MNxtfep8Gt'
            }});
    const response = data as ResponseAxios;
    return response
})
interface initVal {
    movieList: ResponseAxios[];
    genres:string[];
    loading:boolean;
    searchList:ResponseAxios[];
    keyword: string;
    genre:string;
    infoSelect:ResponseAxios;
}
const initialState:initVal = { 
    movieList:[],
    genres:[''],
    loading:false,
    searchList:[],
    keyword:'',
    genre:'',
    infoSelect:{id:0,title:'', genres:['']}
} 
const movie = createSlice({
    name:'movie',
    initialState,
    reducers:{
        filterList:(state,action)=>{
            state.genre = action.payload;
        },
        filterTitle:(state,action)=>{
            state.keyword = action.payload;
          
        },
        selectMovie:(state,action)=>{
            let temp = state.movieList.filter((ele)=>ele.id === action.payload)[0];
            state.infoSelect = temp;
        }
    },
    extraReducers:{
        [hydrateList.pending as any]:(state,action) => {
            state.loading = true;
            console.log('loading...');
        },
        [hydrateList.fulfilled as any]:(state,action) => {
            state.loading = false;
            console.log('complete!');
            const genres:string[] = action.payload.map((ele:ResponseAxios)=>ele.genres).join(',').split(',');

            state.genres = genres.filter((value,index,arr)=>arr.indexOf(value) === index);
            state.movieList = action.payload;
            state.searchList = action.payload;
        },
        [hydrateList.rejected as any]:(state,action) => {
            state.loading = false;
            console.log('failed');
        },
    }
})

export const { filterList, filterTitle, selectMovie } = movie.actions;
export default movie.reducer;