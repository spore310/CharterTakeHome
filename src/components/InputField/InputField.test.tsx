import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { InputField } from './inputField';
import { Provider } from 'react-redux';
import  renderObj  from "../../utils/test-utils"
import { RootState } from '../../redux/store';
import { initialState } from '../../redux/movieSlice';

describe("InputField", () => {

    test("snapshot test", () => {
        const {renderWithProvider} = renderObj;
        const view =renderWithProvider(<InputField />);
        expect(view.asFragment()).toMatchSnapshot();
    });
    test('renders Input Field', () => {
        const {renderWithProvider} = renderObj;
       renderWithProvider(<InputField />);
        const wrapperEl = screen.getByTestId("inputfield-wrapper");
        expect(wrapperEl).toBeInTheDocument();
    });
    test("should render genre list", () => {
        const {renderWithProvider} = renderObj;
        renderWithProvider(<InputField />, {
            preloadedState: {
                movie: {
                    movieList: [],
                    genres: ['Action','Thriller','Crime drama'],
                    loading: false,
                    searchList: [],
                    keyword: '',
                    genre: '',
                    infoSelect: { id: "0", title: '', genres: [''] }
                }
            }
        });
        const genreBtn = screen.getByText("Genre");
        expect(genreBtn).toBeInTheDocument();
        fireEvent.click(genreBtn);
        const ActionOptionEl =  screen.getByText("Action");
        expect(ActionOptionEl).toBeInTheDocument();
    })

})
