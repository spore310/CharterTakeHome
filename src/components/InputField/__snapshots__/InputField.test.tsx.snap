// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField snapshot test 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
    data-testid="inputfield-wrapper"
  >
    <div
      class="inputWrapper"
    >
      <input
        data-testid="inputfield-input"
        type="text"
        value=""
      />
      <div
        class="selectorWrapper"
      >
        <button
          class="genreButton"
        >
          Genre
        </button>
        <div
          class="auto"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
